gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,216 @@
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Queues up the elements until {@link #flush()} is called; once batching is over, returned future
+ * resolves.
+ *
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@InternalApi
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
+    implements Batcher<ElementT, ElementResultT> {
+
+  private final BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT>
+      batchingDescriptor;
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final RequestT prototype;
+  private Batch<ElementT, ElementResultT, RequestT> currentOpenBatch;
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
+  private final Object flushLock = new Object();
+  private boolean isClosed = false;
```

Comment:
Please make it `volatile`

Reply:
This is only used by the single caller thread, so it doesn't need to be volatile

Reply:
Ok, but it is practically free to make it such, ans feels safer (a little bit of defensive programming never hurts in concurrent programming).

Reply:
I generally prefer to avoid defensive programming because you end up hiding the intention of the code. But I agree with you that volatile is free here. So if you feel strongly, then I won't object

Reply:
I would still add volatile

Reply:
@rahulKQL can you add the volatile and I will merge this PR

Reply:
I am sorry, I haven't marked this ongoing conversation as closed.
I have already added a `volatile` for [`isClosed`](https://github.com/googleapis/gax-java/pull/716/files#diff-639d320c75b014154cb1dd32ba32e100R70) flag.